### DISCUSSION TITLE
aligned math, with localized text justifications on the right

### DISCUSSION THREAD

**Discussion by Lakedaemon** [KaTeX/KaTeX#3229] [Q&A]
Hello, 
I am trying to build a system with 
aligned math on the right and ... localized html strings injected to the left as justification for the math steps

Is there a way to do that ?

at the moment, I am using htmlId{} in a \begin{align*}\end{align*} environment,
I render this into a string with katex
AND NEXT, at runtime, in the browser, I am tempering with the html and ids to inject html (for now)
and later compute display metrics and translate the math nodes downards so that the texts do not overlap 

The point is : I compile math nodes once
And next I inject localized (french/english/german text in the html) in the html

![katex](https://user-images.githubusercontent.com/103175/132090653-440670d5-3473-4d06-9719-3b669dedadd4.png)

**Comment by edemaine** [comment]
One natural way to implement this would be to use custom macros whose contents you could pass in to localize.  For example:
```latex
\begin{align*}
f(x) &= \cos(x)^2 & \text{\bydefinition} \\
&= \left(1 - \frac{x^2}2 + \frac{x^4}{4!} + o_0(x^5)\right)^2 & \text{\bylimit} \\
\end{align*}
```
Then at runtime, once you know the desired language, you could call KaTeX with the desired macro definitions:
```js
renderMathInElement(document.body, {
  macros: {
    "\\bydefinition": "par d\\'efinition de la fonction $f$.",
    "\\bylimit": "Via le d\\'eveloppement limit\\'e en $0$ du cosinus \\`a l'ordre $2$.",
  }
});
```

[Here's an example of this running on the demo site.](https://katex.org/?data=%7B%22displayMode%22%3Atrue%2C%22leqno%22%3Afalse%2C%22fleqn%22%3Afalse%2C%22throwOnError%22%3Atrue%2C%22errorColor%22%3A%22%23cc0000%22%2C%22strict%22%3A%22warn%22%2C%22output%22%3A%22htmlAndMathml%22%2C%22trust%22%3Afalse%2C%22macros%22%3A%7B%22%5C%5Cbydefinition%22%3A%22par%20d%5C%5C'efinition%20de%20la%20fonction%20%24f%24.%22%2C%22%5C%5Cbylimit%22%3A%22Via%20le%20d%5C%5C'eveloppement%20limit%5C%5C'e%20en%20%240%24%20du%20cosinus%20%5C%5C%60a%20l'ordre%20%242%24.%22%7D%2C%22code%22%3A%22%5C%5Cbegin%7Balign*%7D%5Cnf(x)%20%26%3D%20%5C%5Ccos(x)%5E2%20%26%20%5C%5Ctext%7B%5C%5Cbydefinition%7D%20%5C%5C%5C%5C%5Cn%26%3D%20%5C%5Cleft(1%20-%20%5C%5Cfrac%7Bx%5E2%7D2%20%2B%20%5C%5Cfrac%7Bx%5E4%7D%7B4!%7D%20%2B%20o_0(x%5E5)%5C%5Cright)%5E2%20%26%20%5C%5Ctext%7B%5C%5Cbylimit%7D%20%5C%5C%5C%5C%5Cn%5C%5Cend%7Balign*%7D%5Cn%22%7D)

**Comment by Lakedaemon** [comment]
Yes, this works when KaTeX dynamically produce HTML from TeX String and a choice of locale

I went another way (for performance reasons) : I store KaTeX MathML/Html rendering of TeX stuff in a database 
and at runtime, I query them and on the user's browser, I stitch these Html strings with localized text to produce the final html that I'm serving to the user 

That way, this is really fast even when I display lots of maths on a page

**Reply by Lakedaemon** [reply]
It makes my life harder and increases the complexity of my app though :/

**Reply by edemaine** [reply]
Ah right, that makes sense. You could use [`\htmlId` or `\htmlClass`](https://katex.org/docs/supported.html#html) to make it easier to find these. (Maybe you already are.) But you still have to fix the horizontal layout, which is a pain.

**Reply by Lakedaemon** [reply]
Yes :)
I'm using HtmlId and HtmlData.
I'll give it a go.
BTW, I really really love KaTeX. Beautifull work there.
